### PR TITLE
Text input with value validation

### PR DESCRIPTION
**Pull request by alansemenov**

### CODE REVIEW

File: modules/lib/src/main/resources/assets/js/app/inputtype/text/CheckedValueInput.ts [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+    private label: LabelEl;
+
+    private inputAndStatusWrapper: DivEl;
+
+    private input: InputEl;
+
+    private statusEl: DivEl;
+
+    private isValueCheckOn: boolean;
```
Comment (alansemenov) [inline modules/lib/src/main/resources/assets/js/app/inputtype/text/CheckedValueInput.ts]:
```suggestion
    private isValueCheckEnabled: boolean;
```

File: modules/lib/src/main/resources/assets/js/app/inputtype/text/CheckedValueInput.ts [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+
+    private debouncedValueHandler: (...args: any[]) => void;
+
+    private stateListeners: { (state: ValueValidationState): void }[] = [];
+
+    private valueCheckStartedListeners: { (): void }[] = [];
```
Comment (alansemenov) [inline modules/lib/src/main/resources/assets/js/app/inputtype/text/CheckedValueInput.ts]:
```suggestion
    private valueCheckListeners: { (): void }[] = [];
```

File: modules/lib/src/main/resources/assets/js/app/inputtype/text/CheckedValueInput.ts [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    }
+
+    private checkValue(): void {
+        const value: string = this.getValue();
+
+        this.validate(value)
+            .then((state: ValueValidationState) => {
+                if (value === this.getValue()) { // value might change during check then no need in update as check is still to be done
+                    this.updateByState(state);
+                    this.setCheckIsOn(false);
```
Comment (alansemenov) [inline modules/lib/src/main/resources/assets/js/app/inputtype/text/CheckedValueInput.ts]:
```suggestion
                    this.setValueCheckEnabled(false);
```

File: modules/lib/src/main/resources/assets/js/app/inputtype/text/CheckedValueInput.ts [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+                }
+            })
+            .catch((e: any) => {
+                this.handlerError();
+                DefaultErrorHandler.handle(e);
+            });
+    }
+
+    protected handlerError(): void {
+        this.setCheckIsOn(false);
```
Comment (alansemenov) [inline modules/lib/src/main/resources/assets/js/app/inputtype/text/CheckedValueInput.ts]:
```suggestion
        this.setValueCheckEnabled(false);
```

File: modules/lib/src/main/resources/assets/js/app/inputtype/text/CheckedValueInput.ts [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+    protected initListeners(): void {
+        this.input.onValueChanged(() => this.handleValueChanged());
+    }
+
+    protected handleValueChanged(): void {
+        if (!this.isValueCheckOn) {
+            this.setCheckIsOn(true);
+            this.notifyValueCheckStarted();
+            this.statusEl.setHtml('');
+        }
```
Comment (alansemenov) [inline modules/lib/src/main/resources/assets/js/app/inputtype/text/CheckedValueInput.ts]:
```suggestion
        if (!this.isValueCheckEnabled) {
            this.setValueCheckEnabled(true);
            this.notifyValueCheck();
            this.statusEl.setHtml('');
        }
```

File: modules/lib/src/main/resources/assets/js/app/inputtype/text/CheckedValueInput.ts [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+        this.initElements();
+        this.initListeners();
+    }
+
+    protected initElements(): void {
+        this.input = new InputEl('input');
+        this.label = new LabelEl(this.getLabelText(), this.input, 'label');
+        this.statusEl = new DivEl('status');
+        this.inputAndStatusWrapper = new DivEl('wrapper');
+        this.debouncedValueHandler = AppHelper.debounce(() => this.checkValue(), this.getValueCheckDelay());
```
Comment (alansemenov) [inline modules/lib/src/main/resources/assets/js/app/inputtype/text/CheckedValueInput.ts]:
```suggestion
        this.debouncedValueCheckHandler = AppHelper.debounce(() => this.checkValue(), this.getValueCheckDelay());
```

File: modules/lib/src/main/resources/assets/js/app/inputtype/text/CheckedValueInput.ts [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+
+    private inputAndStatusWrapper: DivEl;
+
+    private input: InputEl;
+
+    private statusEl: DivEl;
+
+    private isValueCheckOn: boolean;
+
+    private debouncedValueHandler: (...args: any[]) => void;
```
Comment (alansemenov) [inline modules/lib/src/main/resources/assets/js/app/inputtype/text/CheckedValueInput.ts]:
```suggestion
    private debouncedValueCheckHandler: (...args: any[]) => void;
```

File: modules/lib/src/main/resources/assets/js/app/inputtype/text/CheckedValueInput.ts [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+            this.statusEl.setHtml('');
+        }
+
+        this.debouncedValueHandler();
+    }
+
+    private setCheckIsOn(value: boolean): void {
+        this.isValueCheckOn = value;
+        this.statusEl.toggleClass('icon-spinner', value);
+    }
```
Comment (alansemenov) [inline modules/lib/src/main/resources/assets/js/app/inputtype/text/CheckedValueInput.ts]:
```suggestion
    private setValueCheckEnabled(value: boolean): void {
        this.isValueCheckEnabled = value;
        this.statusEl.toggleClass('icon-spinner', value);
    }
```

File: modules/lib/src/main/resources/assets/js/app/inputtype/text/CheckedValueInput.ts [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+
+    protected abstract validate(value: string): Q.Promise<ValueValidationState>;
+
+    protected updateByState(state: ValueValidationState): void {
+        this.updateStatusElement(state);
+        this.notifyState(state);
```
Comment (alansemenov) [inline modules/lib/src/main/resources/assets/js/app/inputtype/text/CheckedValueInput.ts]:
```suggestion
        this.notifyStateUpdate(state);
```

File: modules/lib/src/main/resources/assets/js/app/inputtype/text/CheckedValueInput.ts [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+
+    private input: InputEl;
+
+    private statusEl: DivEl;
+
+    private isValueCheckOn: boolean;
+
+    private debouncedValueHandler: (...args: any[]) => void;
+
+    private stateListeners: { (state: ValueValidationState): void }[] = [];
```
Comment (alansemenov) [inline modules/lib/src/main/resources/assets/js/app/inputtype/text/CheckedValueInput.ts]:
```suggestion
    private stateUpdateListeners: { (state: ValueValidationState): void }[] = [];
```

File: modules/lib/src/main/resources/assets/js/app/inputtype/text/CheckedValueInput.ts [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+        this.stateListeners = this.stateListeners.filter((curr: { (state: ValueValidationState): void }) => {
+            return listener !== curr;
+        });
+    }
+
+    private notifyState(state: ValueValidationState): void {
+        this.stateListeners.forEach((listener: { (state: ValueValidationState): void }) => {
+            listener(state);
+        });
+    }
```
Comment (alansemenov) [inline modules/lib/src/main/resources/assets/js/app/inputtype/text/CheckedValueInput.ts]:
```suggestion
    onStateUpdated(listener: (state: ValueValidationState) => void): void {
        this.stateUpdateListeners.push(listener);
    }

    unStateUpdated(listener: (state: ValueValidationState) => void): void {
        this.stateUpdateListeners = this.stateUpdateListeners.filter((curr: { (state: ValueValidationState): void }) => {
            return listener !== curr;
        });
    }

    private notifyStateUpdated(state: ValueValidationState): void {
        this.stateUpdateListeners.forEach((listener: { (state: ValueValidationState): void }) => {
            listener(state);
        });
    }
```

File: modules/lib/src/main/resources/assets/js/app/inputtype/text/CheckedValueInput.ts [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+        this.valueCheckStartedListeners = this.valueCheckStartedListeners.filter((curr: { (): void }) => {
+            return listener !== curr;
+        });
+    }
+
+    private notifyValueCheckStarted(): void {
+        this.valueCheckStartedListeners.forEach((listener: { (): void }) => {
+            listener();
+        });
+    }
```
Comment (alansemenov) [inline modules/lib/src/main/resources/assets/js/app/inputtype/text/CheckedValueInput.ts]:
```suggestion
    onValueCheck(listener: () => void): void {
        this.valueCheckListeners.push(listener);
    }

    unValueCheck(listener: () => void): void {
        this.valueCheckListeners = this.valueCheckListeners.filter((curr: { (): void }) => {
            return listener !== curr;
        });
    }

    private notifyValueCheck(): void {
        this.valueCheckListeners.forEach((listener: { (): void }) => {
            listener();
        });
    }
```

File: modules/lib/src/main/resources/assets/js/app/inputtype/text/CheckedValueInput.ts (hunk 8 above)
Comment (ashklianko) [reply modules/lib/src/main/resources/assets/js/app/inputtype/text/CheckedValueInput.ts]: `setValueCheckEnabled` is a misleading name, we're trying to say that the check is now in progress, we do not let disable or enable check by demand, so `setCheckIsInProgress` would be better then

Comment (alansemenov) [reply modules/lib/src/main/resources/assets/js/app/inputtype/text/CheckedValueInput.ts]: Let's go for `isCheckInProgress` and `setIsCheckInProgress` then.

**Pull request merged**